compiler/rustc_infer/src/infer/error_reporting/note_and_explain.rs

```diff
@@ -32,6 +32,12 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                         diag.note("no two closures, even if identical, have the same type");
                         diag.help("consider boxing your closure and/or using it as a trait object");
                     }
+                    (ty::Coroutine(def_id1, ..), ty::Coroutine(def_id2, ..))
+                        if self.tcx.coroutine_is_async(def_id1)
+                            && self.tcx.coroutine_is_async(def_id2) =>
+                    {
+                        diag.note("every async block has a different type even when the output type is the same");
```

Comment:
```suggestion
                        diag.note("no two async blocks, even if identical, have the same type");
                        diag.help("consider pinning your async block and and casting it to a trait object");
```
copying from the suggestion for closures. Looking at the test, a fixed version would be something like
```rust
fn fun<F: Future<Output = ()>>(f1: F, f2: F) {}
fn main() {
    fun(Box::pin(async {}) as Pin<Box<dyn Future<Output = ()>>>, Box::pin(async {}));
}
```

Reply:
Done